### PR TITLE
feat(ST-40): add create classroom api endpoint

### PR DESCRIPTION
**Pull request by faysal-alam-sazim**
## Description of Change

- Added POST /classrooms endpoint to create classroom in the classroom controller
- Added addClassroom function in the service layer to create a Classroom and Teacher Object
- In the repository layer, persist and flush is happening

## Associated Ticket Link(s)

- https://trello.com/c/BFr3px4u

## Related Pull Request(s)

- N/A

### CODE REVIEW

File: src/classrooms/classrooms.controller.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  user: ITokenizedUser;
+}
+
+@UseGuards(JwtAuthGuard)
+@Controller("classrooms")
+export class ClassroomsController {
+  constructor(private readonly classroomsService: ClassroomsService) {}
+
+  @Post()
+  addClassroom(@Req() req: IRequestUser, @Body() classroomDto: CreateClassroomDto) {
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts:19]:
```suggestion
  createClassroom(@Req() req: IRequestUser, @Body() classroomDto: CreateClassroomDto) {
```

Use @CurrentUser decorator if you need current user id. Can find it in the backend template

File: src/classrooms/classrooms.module.ts [hunk 2]
```diff
@@ -3,6 +3,19 @@ import { Module } from "@nestjs/common";
...
 
-@Module({ imports: [MikroOrmModule.forFeature([Classroom])] })
+import { ClassroomsController } from "./classrooms.controller";
+import { ClassroomsService } from "./classrooms.service";
+
+@Module({
+  imports: [
+    MikroOrmModule.forFeature([Classroom]),
+    MikroOrmModule.forFeature([User]),
+    MikroOrmModule.forFeature([Teacher]),
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.module.ts:16]:
```suggestion
    MikroOrmModule.forFeature([Classroom, User, Teacher]),
```

File: src/classrooms/classrooms.repository.ts [hunk 3]
```diff
@@ -3,6 +3,13 @@ import { Injectable } from "@nestjs/common";
...
+import { Teacher } from "@/common/entities/teachers.entity";
 
 @Injectable()
-export class ClassroomRepository extends EntityRepository<Classroom> {}
+export class ClassroomRepository extends EntityRepository<Classroom> {
+  async addClassroom(classroom: Classroom, teacher: Teacher) {
+    this.em.persist([classroom, teacher]);
+    await this.em.flush();
+    return classroom;
+  }
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.repository.ts:14]: Don't need this method. Use createOne in service and persist and flush there

File: src/classrooms/classrooms.service.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+export class ClassroomsService {
+  constructor(
+    private readonly classroomRepository: ClassroomRepository,
+    private readonly teacherRepository: TeachersRepository,
+    private readonly userRepository: UsersRepository,
+  ) {}
+
+  async addClassroom(classroomDto: CreateClassroomDto, userId: number) {
+    const { title, subject, classTime, days } = classroomDto;
+    const classroom = new Classroom(title, subject, classTime, days);
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts:20]: No need to destructure. Use repository createOne method and pass the dto.

File: src/db/migrations/Migration20240829131904_classroom.ts [hunk 5]
```diff
@@ -1,15 +1,17 @@
-import { Migration } from '@mikro-orm/migrations';
+import { Migration } from "@mikro-orm/migrations";
```
Comment (ishmam-iqbal-sazim) [inline src/db/migrations/Migration20240829131904_classroom.ts:1]: These migrations were in the previous PR. Shouldn't see these here

File: src/db/migrations/Migration20240829144904_classroom_field_name_update.ts [hunk 6]
```diff
@@ -1,7 +1,6 @@
-import { Migration } from '@mikro-orm/migrations';
+import { Migration } from "@mikro-orm/migrations";
```
Comment (ishmam-iqbal-sazim) [inline src/db/migrations/Migration20240829144904_classroom_field_name_update.ts:1]: Same comment as above

File: src/classrooms/classrooms.controller.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+import { CreateClassroomDto } from "./classrooms.dto";
+import { ClassroomsService } from "./classrooms.service";
+
+interface IRequestUser extends Request {
+  user: ITokenizedUser;
+}
+
+@UseGuards(JwtAuthGuard)
+@Controller("classrooms")
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.controller.ts:14]: Shouldn't only teachers be able to create classroom?

File: src/classrooms/classrooms.controller.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import { ClassroomsService } from "./classrooms.service";
+
+@UseGuards(JwtAuthGuard)
+@Controller("classrooms")
+export class ClassroomsController {
+  constructor(private readonly classroomsService: ClassroomsService) {}
+
+  @Post()
+  createClassroom(@CurrentUser() user: ITokenizedUser, @Body() classroomDto: CreateClassroomDto) {
+    if (user.claim === EUserRole.STUDENT) {
```
Comment (ragib-hossain-sazim) [inline src/classrooms/classrooms.controller.ts:18]: @ishmam-iqbal-sazim is this correct way to guard endpoints? Shouldn't there be a roles guard decorator instead?

Comment (ishmam-iqbal-sazim) [reply src/classrooms/classrooms.controller.ts:18]: No, it is best to use a guard with a roles decorator with a guard. Implementation can be found in the backend template

File: src/classrooms/classrooms.service.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import { UsersRepository } from "@/users/users.repository";
+
+import { CreateClassroomDto } from "./classrooms.dto";
+import { ClassroomsRepository } from "./classrooms.repository";
+
+@Injectable()
+export class ClassroomsService {
+  constructor(
+    private readonly classroomRepository: ClassroomsRepository,
+    private readonly userRepository: UsersRepository,
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts:14]:
```suggestion
    private readonly classroomsRepository: ClassroomsRepository,
    private readonly usersRepository: UsersRepository,
```
[NIT]

File: src/classrooms/classrooms.service.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  ) {}
+
+  async createClassroom(classroomDto: CreateClassroomDto, userId: number) {
+    const user = await this.userRepository.findOneOrFail({ id: userId });
+    const classroom = this.classroomRepository.create({
+      ...classroomDto,
+      teacher: user.teacher,
+    });
+    await this.em.persistAndFlush(classroom);
+    return classroom;
```
Comment (ishmam-iqbal-sazim) [inline src/classrooms/classrooms.service.ts:25]:
```suggestion
    const user = await this.userRepository.findOneOrFail({ id: userId });

    const classroom = this.classroomRepository.create({
      ...classroomDto,
      teacher: user.teacher,
    });

    await this.em.persistAndFlush(classroom);

    return classroom;
```

Let's add line breaks here

**Review by ishmam-iqbal-sazim — APPROVED** [review]

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**